[@kbn/scout] Add Scout API `expect` with custom matchers

steliosmavro opened the pull request:
# Scout API Matchers

## Summary

Scout exposes a custom API `expect` object with custom and default matchers for API testing. To enforce best practices, Playwright's default matchers are restricted. Also updates all API tests in `@kbn/scout` to use the new matchers.

```typescript
import { expect } from '@kbn/scout/api';
```

---

## Custom Response Matchers

Scout-specific matchers for API response assertions. These support both `apiClient` and `kbnClient` response interfaces.

### toHaveStatusCode

Asserts response has the expected HTTP status code. Checks `status` or `statusCode` property.

```typescript
expect(response).toHaveStatusCode(200);
expect(response).toHaveStatusCode({ oneOf: [200, 201] });
```

### toHaveStatusText

Asserts response has the expected status text. Checks `statusText` or `statusMessage` property.

```typescript
expect(response).toHaveStatusText('OK');
```

### toHaveHeaders

Asserts response contains the expected headers (partial match). Checks `headers` property. Case-insensitive keys. Response header arrays are joined with `, ` before comparison.

```typescript
expect(response).toHaveHeaders({ 'content-type': 'application/json' });
expect(response).toHaveHeaders({ 'set-cookie': 'session=abc, token=xyz' });
```

---

## Custom Asymmetric Matchers

Custom asymmetric matchers for use with `toMatchObject`:

- **`expect.toBeGreaterThan(n)`** - Matches if value > n
- **`expect.toBeLessThan(n)`** - Matches if value < n

```typescript
expect(response).toMatchObject({
  body: {
    count: expect.toBeGreaterThan(0),
    limit: expect.toBeLessThan(100),
  },
});
```

---

## Supported Playwright Matchers

The following [Playwright matchers](https://playwright.dev/docs/api/class-genericassertions) are available:

**Matchers:**

- `toBe(expected)` - compares using `Object.is`, use for primitive literals
- `toBeDefined()` - value is not `undefined`
- `toBeUndefined()` - value is `undefined`
- `toContain(expected)` - string contains substring, or Array/Set contains item
- `toHaveLength(n)` - value has `.length` property equal to n
- `toStrictEqual(expected)` - deep equality with type checking
- `toBeGreaterThan(n)` - value > n
- `toBeLessThan(n)` - value < n
- `toMatchObject(expected)` - partial object matching

**Asymmetric matchers:**

- `expect.arrayContaining(array)` - array contains all expected elements
- `expect.objectContaining(object)` - object contains expected properties

---

## Restricted Matchers

To enforce best practices, some Playwright matchers are restricted. Use these alternatives:

```typescript
expect({ name: 'Alice', metadata: undefined }).toEqual({ name: 'Alice' }); // ❌ would pass even if extra keys exist
expect({ name: 'Alice', metadata: undefined }).toStrictEqual({ name: 'Alice' }); // ✅ fails if extra/missing keys exist

expect(response).toHaveProperty('apiKey'); // ❌ passes even if apiKey is undefined
expect(response.apiKey).toBeDefined(); // ✅ it will fail if undefined

expect(exists).toBeTruthy(); // ❌ too loose
expect(exists).toBe(true); // ✅ be explicit

expect(response.apiKey).not.toBeDefined(); // ❌ not available
expect(response.apiKey).toBeUndefined(); // ✅ preferred by playwright
```

**UI-specific matchers** like `toBeVisible`, `toBeEnabled`, `toHaveAttribute`, etc. are not available for API tests.

## Notes

- Custom matchers throw Playwright-style errors
- Stack traces can be trimmed via `skipStackLines` to point directly at the test file
- `toHaveStatusCode` options interface is designed for extensibility (e.g., future `range` option)


<!--ONMERGE {"backportTargets":["8.19","9.1","9.2","9.3"]} ONMERGE-->

steliosmavro committed:
[@kbn/scout] Extend Playwright matchers for API tests

elasticmachine commented:
<!-- buildkite-pr-bot-trigger-comment -->:robot: Jobs for this PR can be triggered through checkboxes. :construction: 

:information_source: To trigger the CI, please tick the checkbox below :point_down: 

- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-pull-request --> Click to trigger **kibana-pull-request** for this PR!
- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-deploy-project-from-pr --> Click to trigger **kibana-deploy-project-from-pr** for this PR!
- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-deploy-cloud-from-pr --> Click to trigger **kibana-deploy-cloud-from-pr** for this PR!
- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-entity-store-performance-from-pr --> Click to trigger **kibana-entity-store-performance-from-pr** for this PR!

steliosmavro committed:
Add API matchers for api-response and generic assertions

steliosmavro committed:
Add `toHaveData` matcher, simplify tests to not check error msg, and update custom matchers to use playwright's expect under the hood

steliosmavro committed:
Add dynamic matchers

steliosmavro committed:
Refactor

steliosmavro committed:
Enhance `toHaveData` matcher with partial matching support for arrays and nested structures

steliosmavro committed:
Remove `createApiResponse` since every matcher now only looks for its own property

steliosmavro committed:
Add `oneOf` option for `toHaveStatusCode`. Update scout tests to use custom API matchers exclusively (except `toBeDefined`). Comment out all unused `ValueMatchers`.

steliosmavro committed:
Clean up

steliosmavro committed:
Add README.md

steliosmavro committed:
minor update README.md

steliosmavro committed:
Merge branch 'main' into extend-scout-matchers-for-api-tests

steliosmavro committed:
Add asymmetric matchers

steliosmavro committed:
Add support for `apiClient` response. Rename to `toHavePayload`.

steliosmavro committed:
Add skipping stack lines option when throwing matcher error

steliosmavro committed:
Update README.md file

steliosmavro committed:
Updated matchers:
- Split `toHavePayload` into `toHaveBody` and `toHaveData`.
- Added value matchers `toStrictEqual`, toBeGreaterThan, toBeLessThan.
- Removed `toHaveLength` asymmetric matcher.
- Updated all API test specs and documentation accordingly.

steliosmavro committed:
Merge branch 'main' into extend-scout-matchers-for-api-tests

csr commented on src/platform/packages/shared/kbn-scout/api.ts:
```diff
@@ -0,0 +1,10 @@
+ * License v3.0 only", or the "Server Side Public License, v 1".
+ */
+
+export { expect } from './src/playwright/matchers/api';
```
Are we planning to make this `expect` available in the other Scout solution packages? At the moment, AFAIK, the custom API matchers are not exposed via solution‑specific Scout packages (e.g. `@kbn/scout-oblt`), so they can't be used in solution-specific tests.

steliosmavro replied:
Great point. I'm not 100% sure but to give you my opinion about it, api matchers are not like fixtures. Fixtures are encouraged to be custom made for specific use cases. Matchers are not. Besides I doubt that they will need solution specific matchers. Again, it's not like fixtures. But this is my understanding, I could be way wrong. 😊

csr reviewed (commented):
Nice work! I left a few comments, which you're free to address at your discretion.

steliosmavro committed:
Update src/platform/packages/shared/kbn-scout/src/playwright/matchers/api/README.md

Co-authored-by: Cesare de Cal <cesare.decal@elastic.co>

steliosmavro commented on src/platform/packages/shared/kbn-scout/src/playwright/matchers/api/README.md:
```diff
@@ -0,0 +1,181 @@
+
+## Summary
+
+Create a Scout API custom `expect` with dynamic matchers for API testing and restricted exposure to Playwright's default matchers. This is to assist developers with API-specific assertions and enforce best practices
```
I like it, thanks @csr !

steliosmavro committed:
Remove `toHavePayload` matcher and `dynamic_matchers.ts`

- Add generic matchers: `toBe`, `toBeUndefined`, `toContain`, `toHaveLength`
- Add custom asymmetric matchers: `toBeGreaterThan`, `toBeLessThan`
- Derive `AsymmetricMatcher` type from Playwright instead of custom interface
- Rename `value_matchers.ts` to `generic_matchers.ts`
- Add `response_matchers.ts` for response-specific matchers
- Update README with supported/restricted matchers documentation
- Update API test specs to use new matchers

steliosmavro committed:
Merge branch 'main' into extend-scout-matchers-for-api-tests

steliosmavro committed:
Merge branch 'main' into extend-scout-matchers-for-api-tests

steliosmavro committed:
Remove `not` support from response matchers in Scout API expect and from `toBeDefined` generic matcher

steliosmavro committed:
Merge branch 'main' into extend-scout-matchers-for-api-tests

dmlemeshko reviewed (approved):
LGTM. I think it is a great set of API matchers to start with it.

Thank you @steliosmavro for iterating though all the comments / requests and polishing it 👏
